In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

# Importing the provided CSV file.

In [2]:
df1 = pd.read_csv("./news.csv")
df1.head()

title  \
0  BTS: RM is reminded of Bon Voyage as he travel...   
1  RM recalls wondering if he 'made right decisio...   
2  BTS: J-Hope and RM go bonkers at Billie Eilish...   
3  BTS: J-Hope proudly states he raised Jungkook,...   
4  BTS: Jin aka Kim Seokjin takes us through the ...   

                                             content          published_at  \
0  After reaching his hotel in the city, RM revea...  2022-07-30T07:00:00Z   
1  RM aka Kim Namjoon was the first member to joi...  2022-12-22T15:57:55Z   
2  Billie Eilish's concert was held in Seoul, Sou...  2022-08-16T07:00:00Z   
3  BTS ARMY y'all would be missing the members a ...  2022-12-18T13:08:40Z   
4  BTS member Kim Seokjin aka Jin has the capacit...  2022-11-21T08:00:00Z   

   source  topic  
0       2     13  
1       2     13  
2       1      7  
3       1      7  
4       1      8

In [3]:
df1.shape

(810, 5)

In [4]:
df1['content'].iloc[401]

'BTS consists of members Jin, Suga, J-Hope, RM, Jimin, V, and Jungkook&mdash;co-writes and co-produces much of their own material. Originally a hip hop group, their musical style has evolved to incorporate a wide range of genres; their lyrics have often discussed mental health, the troubles of school-age youth and coming of age, loss, the journey towards self-love, and individualism. Their work also frequently references literature, philosophy and psychological concepts, and includes an alternate universe storyline.\n\nBTS in the beginning:\n\nAfter launching in 2013 with their single album 2 Cool 4 Skool, BTS respectively released their first Korean-language studio album, Dark &amp; Wild, and Japanese-language studio album, Wake Up, in 2014. The group&#39;s second Korean studio album, Wings (2016), was their first to sell one million copies in South Korea. By 2017, BTS had crossed into the global music market, leading the Korean wave into the United States and breaking several sales r

In [5]:
df1.drop(401, axis=0, inplace=True)

In [6]:
df1.reset_index()
df1['content'].iloc[404]

'J-Hope of BTS, earlier released his debut solo album, Jack in the Box, featuring 10 tracks including the lead single &#39;More&#39; and the follow-up &#39;Arson.&#39; Both singles received music videos, and the album was released on July 15th after BTS announced a temporary break to allow for solo pursuits.\n\nJ-Hope- &lsquo;Jack in the Box&rsquo;\n\nJ-Hope&#39;s solo album, Jack in the Box, is a concept album exploring themes of passion, ambition, insecurity, success, and anxiety. The album&#39;s central concept revolves around Pandora&#39;s box, and the music itself is a blend of old-school hip hop and elements of pop, grunge, and R&amp;B. Fans and critics alike have praised the album for its thought-provoking lyrics and diverse sound. J-Hope&#39;s solo debut has been a success, showcasing his unique style and versatility as an artist. Here are his top 5 solo performances of 2022:&nbsp;\n\n1. J-Hope at Lollapalooza music festival in Chicago\n6ak5yOmj8W8J-Hope made history as the fir

### Realizing that other columns other than published_at, source, topic are useless.

In [7]:
df1.drop(['published_at', 'source', 'topic'], axis=1)

title  \
0    BTS: RM is reminded of Bon Voyage as he travel...   
1    RM recalls wondering if he 'made right decisio...   
2    BTS: J-Hope and RM go bonkers at Billie Eilish...   
3    BTS: J-Hope proudly states he raised Jungkook,...   
4    BTS: Jin aka Kim Seokjin takes us through the ...   
..                                                 ...   
805  BTS’ SUGA’s Suchwita Ep 2 Teaser OUT: Top 3 so...   
806  BTS ARMY celebrate 700 days of Jin's special s...   
807  BTS: 'I am not a baby,' says Jungkook as an AR...   
808  BTS' Jin shares 1st pics after joining militar...   
809  Bad Decisions: BTS’ Jin, Jimin, V & Jungkook j...   

                                               content  
0    After reaching his hotel in the city, RM revea...  
1    RM aka Kim Namjoon was the first member to joi...  
2    Billie Eilish's concert was held in Seoul, Sou...  
3    BTS ARMY y'all would be missing the members a ...  
4    BTS member Kim Seokjin aka Jin has the capacit...  
..                                                 ...  
805  BTS has conquered the world with their group r...  
806  Today marks 700 days since BTS' worldwide hand...  
807  BTS' youngest member Jungkook came online on W...  
808  BTS' eldest member Jin has shared pictures and...  
809  After a lot of teasing, Benny Blanco’s collabo...  

[809 rows x 2 columns]

In [8]:
df1.drop(['published_at', 'source', 'topic'], axis=1, inplace=True)
df1.head()

title  \
0  BTS: RM is reminded of Bon Voyage as he travel...   
1  RM recalls wondering if he 'made right decisio...   
2  BTS: J-Hope and RM go bonkers at Billie Eilish...   
3  BTS: J-Hope proudly states he raised Jungkook,...   
4  BTS: Jin aka Kim Seokjin takes us through the ...   

                                             content  
0  After reaching his hotel in the city, RM revea...  
1  RM aka Kim Namjoon was the first member to joi...  
2  Billie Eilish's concert was held in Seoul, Sou...  
3  BTS ARMY y'all would be missing the members a ...  
4  BTS member Kim Seokjin aka Jin has the capacit...

### Checking if there are any null values in the dataframe.

In [9]:
df1.isna().sum()

title      0
content    4
dtype: int64

### As there are null values and are very less (4 <<< 810), its better to drop them.

In [10]:
df1.dropna(axis=0, inplace=True)
df1.head()

title  \
0  BTS: RM is reminded of Bon Voyage as he travel...   
1  RM recalls wondering if he 'made right decisio...   
2  BTS: J-Hope and RM go bonkers at Billie Eilish...   
3  BTS: J-Hope proudly states he raised Jungkook,...   
4  BTS: Jin aka Kim Seokjin takes us through the ...   

                                             content  
0  After reaching his hotel in the city, RM revea...  
1  RM aka Kim Namjoon was the first member to joi...  
2  Billie Eilish's concert was held in Seoul, Sou...  
3  BTS ARMY y'all would be missing the members a ...  
4  BTS member Kim Seokjin aka Jin has the capacit...

In [11]:
df1.shape

(805, 2)

# Discovering irregularities in data

In [12]:
df1['content'].iloc[0]

'After reaching his hotel in the city, RM revealed that his stay would be for four days and added that he would step out for dinner. As he sat at a roadside open-air restaurant, RM feasted on beer, burgers and fries. He said, "I\'m starving right now. I\'m out to grab some food. It\'s much quieter than I expected and feels like a rural town. I like the familiar atmosphere." RM attended Art Basel and explained on camera the details of the art fair. He also gave a glimpse as he had noodles and beer which was followed by soup noodles and wrap. Showing the pattern of a ping pong table, RM said, "The table looks like our (BTS) symbol." He also spoke about the art pieces as he viewed them. After that, RM took a tram to visit the Foundation Beyeler, a museum. He later took a walk through the city. On his third day, RM visited the Kunstmuseum Basel, the Vitra Design Museum and the gallery. As he walked around, RM showed a chair to his fans and said, "I have breaking news for you guys. Coldplay

In [13]:
df1['content'].iloc[365]

'BTS&rsquo; Jin is good at a lot of things, singing, being &lsquo;worldwide handsome&rsquo;, cooking for Bangtan, and now even making alcohol  The second week of his own program, Drunken Truth, which starred well-known chef Baek Jong Won and saw a guest appearance from actor Kim Nam Gil, unfolds in unique ways \n\nDrunken Truth Episode 3\nAs Jin and Baek Jong Won reach a spot to take care of the final steps of his first try at making makgeolli (rice wine), the two show another attempt at their flourishing synergy  Jin is known to be good with elders as he can joke around and that is clearly visible in the show  The two head out with their own bottles of self-made alcohol, their destination is a traditional market  Being a chef, it is Baek Jong Won&rsquo;s territory who is recognized by the vendors which consist of mostly older women and is asked for photos, while funnily enough, they are unaware of BTS member Jin  A neck-and-neck competition between them shows the difference in prefere

In [14]:
text = df1['content'].iloc[369]
text

'2022 came to a celebratory end for BTS, especially member J-Hope, who was in New York, making another fabulous display of his skills  While away from the members, he seemed to have enjoyed it to the fullest with a solo performance at Dick Clark&rsquo;s New Year&rsquo;s Rockin&rsquo; Eve where he did a live stage of 3 songs, solo track &lsquo;= (Equal Sign)&rsquo;, &lsquo;Chicken Noodle Soup&rsquo; his collaboration track with Becky G, and BTS&rsquo; &lsquo;Butter&rsquo; (Holiday Remix)  He officially became only the second South Korean soloist to perform at the event, following PSY  This was also J-Hope&rsquo;s third time at Dick Clark&rsquo;s New Year&rsquo;s Rockin&rsquo; Eve after group stages with BTS in 2017 and 2019 \n\nJ-Hope&rsquo;s live\n\nAfter returning from Times Square where he performed as the penultimate act alongside multiple other singers from around the world and spoke to Ryan Seacrest, the host of the show, J-Hope was back at his hotel room and turned on a live broa

In [15]:
df1['content'].iloc[450]

'On May 27 at 8:30 pm IST (May 28 at midnight KST), BTS dropped the first set of concept photos for their upcoming anthology album, ‘Proof’  The stunning photos feature the members standing in front of the door of a vault, decked out in all-black outfits  Meanwhile, the individual photos of the members show them standing in front of a shield that is riddled with bullets, similar to the door of the vault, showing signs of wear and tear, but still standing strong \xa0Check out the first set of concept photos for BTS’ ‘Proof’, below:https://twitter com/BIGHIT_MUSIC/status/1530202174600675328?ref_src=twsrc%5EtfwAs BTS is an acronym for their Korean name, ‘Bangtan Sonyeondan’, which translates to ‘Bulletproof Boys’ / ‘Bulletproof Boy Scouts’, this concept takes us back to the very beginning of BTS  The group’s name represents their desire to “block out stereotypes, criticisms, and expectations that aim on adolescents like bullets ”\xa0With the group’s upcoming anthology set to “reflect the 

### There are HTML Tags and Entities, which needs to be cleaned from data.

In [16]:
from bs4 import BeautifulSoup

In [17]:
def removeHTMLTagsAndEntities(s):
    return BeautifulSoup(BeautifulSoup(s, "lxml").text, "html.parser")

In [18]:
removeHTMLTagsAndEntities(df1['content'].iloc[365])

BTS’ Jin is good at a lot of things, singing, being ‘worldwide handsome’, cooking for Bangtan, and now even making alcohol  The second week of his own program, Drunken Truth, which starred well-known chef Baek Jong Won and saw a guest appearance from actor Kim Nam Gil, unfolds in unique ways 

Drunken Truth Episode 3
As Jin and Baek Jong Won reach a spot to take care of the final steps of his first try at making makgeolli (rice wine), the two show another attempt at their flourishing synergy  Jin is known to be good with elders as he can joke around and that is clearly visible in the show  The two head out with their own bottles of self-made alcohol, their destination is a traditional market  Being a chef, it is Baek Jong Won’s territory who is recognized by the vendors which consist of mostly older women and is asked for photos, while funnily enough, they are unaware of BTS member Jin  A neck-and-neck competition between them shows the difference in preferences between the older and y

In [19]:
text = removeHTMLTagsAndEntities(text)
text

2022 came to a celebratory end for BTS, especially member J-Hope, who was in New York, making another fabulous display of his skills  While away from the members, he seemed to have enjoyed it to the fullest with a solo performance at Dick Clark’s New Year’s Rockin’ Eve where he did a live stage of 3 songs, solo track ‘= (Equal Sign)’, ‘Chicken Noodle Soup’ his collaboration track with Becky G, and BTS’ ‘Butter’ (Holiday Remix)  He officially became only the second South Korean soloist to perform at the event, following PSY  This was also J-Hope’s third time at Dick Clark’s New Year’s Rockin’ Eve after group stages with BTS in 2017 and 2019 

J-Hope’s live

After returning from Times Square where he performed as the penultimate act alongside multiple other singers from around the world and spoke to Ryan Seacrest, the host of the show, J-Hope was back at his hotel room and turned on a live broadcast to speak with his fans  As they congratulated him on his sparkling performance, being the

### There are many spaces, new line characters and also some sentences are not even having full stops.

In [20]:
import re

In [21]:
def addingFullstops(s):
    s = re.sub("\xa0",' ', str(s))
    s = re.sub("(\\n)+", '.', str(s))
    s = re.sub("(\\t)+", '', str(s))
    s = re.sub("   ", '', str(s))
    s = re.sub("(  )",'.', str(s))
    return s

In [22]:
text = addingFullstops(text)
text

'2022 came to a celebratory end for BTS, especially member J-Hope, who was in New York, making another fabulous display of his skills.While away from the members, he seemed to have enjoyed it to the fullest with a solo performance at Dick Clark’s New Year’s Rockin’ Eve where he did a live stage of 3 songs, solo track ‘= (Equal Sign)’, ‘Chicken Noodle Soup’ his collaboration track with Becky G, and BTS’ ‘Butter’ (Holiday Remix).He officially became only the second South Korean soloist to perform at the event, following PSY.This was also J-Hope’s third time at Dick Clark’s New Year’s Rockin’ Eve after group stages with BTS in 2017 and 2019 .J-Hope’s live.After returning from Times Square where he performed as the penultimate act alongside multiple other singers from around the world and spoke to Ryan Seacrest, the host of the show, J-Hope was back at his hotel room and turned on a live broadcast to speak with his fans.As they congratulated him on his sparkling performance, being the perf

### Even after cleaning, there are some repeated full stops without any sentence. Pre-processing it

In [23]:
from nltk.tokenize import sent_tokenize

In [24]:
def addFullStops(s):
    modified_sentences = []
    sentences = s.split(".")
    for i in sentences:
        i = i.strip()
        if len(i) != 0:
            i += '.'
            modified_sentences.append(i)
    return " ". join(modified_sentences)

In [25]:
text = addFullStops(text)
text

'2022 came to a celebratory end for BTS, especially member J-Hope, who was in New York, making another fabulous display of his skills. While away from the members, he seemed to have enjoyed it to the fullest with a solo performance at Dick Clark’s New Year’s Rockin’ Eve where he did a live stage of 3 songs, solo track ‘= (Equal Sign)’, ‘Chicken Noodle Soup’ his collaboration track with Becky G, and BTS’ ‘Butter’ (Holiday Remix). He officially became only the second South Korean soloist to perform at the event, following PSY. This was also J-Hope’s third time at Dick Clark’s New Year’s Rockin’ Eve after group stages with BTS in 2017 and 2019. J-Hope’s live. After returning from Times Square where he performed as the penultimate act alongside multiple other singers from around the world and spoke to Ryan Seacrest, the host of the show, J-Hope was back at his hotel room and turned on a live broadcast to speak with his fans. As they congratulated him on his sparkling performance, being the

### For certain words, I saw contractions were used, so decided to replace them with actual word.

In [26]:
import contractions

In [27]:
def removingContractions(s):
    words = []
    for i in s.split(" "):
        words.append(contractions.fix(i))
    s = " ".join(words)
    return s

In [28]:
removingContractions("I'd")

'I would'

In [29]:
text = removingContractions(text)
text

'2022 came to a celebratory end for BTS, especially member J-Hope, who was in New York, making another fabulous display of his skills. While away from the members, he seemed to have enjoyed it to the fullest with a solo performance at Dick Clark’s New Year’s Rockin’ Eve where he did a live stage of 3 songs, solo track ‘= (Equal Sign)’, ‘Chicken Noodle Soup’ his collaboration track with Becky G, and BTS’ ‘Butter’ (Holiday Remix). He officially became only the second South Korean soloist to perform at the event, following PSY. This was also J-Hope’s third time at Dick Clark’s New Year’s Rockin’ Eve after group stages with BTS in 2017 and 2019. J-Hope’s live. After returning from Times Square where he performed as the penultimate act alongside multiple other singers from around the world and spoke to Ryan Seacrest, the host of the show, J-Hope was back at his hotel room and turned on a live broadcast to speak with his fans. As they congratulated him on his sparkling performance, being the

### Till here combining what we did

In [30]:
def initial_preprocessing(s):
    s = removeHTMLTagsAndEntities(s)
    s = addingFullstops(s)
    s = removingContractions(s)
    return s

In [31]:
initial_preprocessing(df1['content'].iloc[369])

'2022 came to a celebratory end for BTS, especially member J-Hope, who was in New York, making another fabulous display of his skills.While away from the members, he seemed to have enjoyed it to the fullest with a solo performance at Dick Clark’s New Year’s Rockin’ Eve where he did a live stage of 3 songs, solo track ‘= (Equal Sign)’, ‘Chicken Noodle Soup’ his collaboration track with Becky G, and BTS’ ‘Butter’ (Holiday Remix).He officially became only the second South Korean soloist to perform at the event, following PSY.This was also J-Hope’s third time at Dick Clark’s New Year’s Rockin’ Eve after group stages with BTS in 2017 and 2019 .J-Hope’s live.After returning from Times Square where he performed as the penultimate act alongside multiple other singers from around the world and spoke to Ryan Seacrest, the host of the show, J-Hope was back at his hotel room and turned on a live broadcast to speak with his fans.As they congratulated him on his sparkling performance, being the perf

In [32]:
initial_preprocessing(df1['content'].iloc[0])

'After reaching his hotel in the city, RM revealed that his stay would be for four days and added that he would step out for dinner. As he sat at a roadside open-air restaurant, RM feasted on beer, burgers and fries. He said, "I am starving right now. I am out to grab some food. It is much quieter than I expected and feels like a rural town. I like the familiar atmosphere." RM attended Art Basel and explained on camera the details of the art fair. He also gave a glimpse as he had noodles and beer which was followed by soup noodles and wrap. Showing the pattern of a ping pong table, RM said, "The table looks like our (BTS) symbol." He also spoke about the art pieces as he viewed them. After that, RM took a tram to visit the Foundation Beyeler, a museum. He later took a walk through the city. On his third day, RM visited the Kunstmuseum Basel, the Vitra Design Museum and the gallery. As he walked around, RM showed a chair to his fans and said, "I have breaking news for you guys. Coldplay

### Removing the punctuations and Stopwords and coverting entire sentence into lower case.

In [33]:
from nltk.corpus import stopwords

In [34]:
engStopwords = stopwords.words("english")
print(engStopwords)

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

In [35]:
def removePunctuations(s):
    words = []
    s = re.sub(r"[<>()|&©ø\[\]\'\",;?~*!]", ' ', str(s))
    for i in s.split(" "):
        if i not in engStopwords:
            words.append(i)
    s = " ".join(words)
    return s

In [36]:
removePunctuations(text)

'2022 came celebratory end BTS  especially member J-Hope  New York  making another fabulous display skills. While away members  seemed enjoyed fullest solo performance Dick Clark’s New Year’s Rockin’ Eve live stage 3 songs  solo track ‘=  Equal Sign ’  ‘Chicken Noodle Soup’ collaboration track Becky G  BTS’ ‘Butter’  Holiday Remix . He officially became second South Korean soloist perform event  following PSY. This also J-Hope’s third time Dick Clark’s New Year’s Rockin’ Eve group stages BTS 2017 2019. J-Hope’s live. After returning Times Square performed penultimate act alongside multiple singers around world spoke Ryan Seacrest  host show  J-Hope back hotel room turned live broadcast speak fans. As congratulated sparkling performance  perfectionist  J-Hope expressed sadness unable use voice fullest talked slipping incident due rainy weather rehearsal stage. J-Hope BTS members’ New Year wishes. As soon turned 12  New Year began South Korea  BTS members Jungkook  Jimin  V took fan comm

# Trying to combine all of these pre-processing in a single function 

In [37]:
def preprocessingRow(s):
    s = BeautifulSoup(BeautifulSoup(s, "lxml").text, "html.parser")
    s = re.sub("\xa0",' ', str(s))
    s = re.sub("(\\n)+", '.', str(s))
    s = re.sub("(\\t)+", '', str(s))
    s = re.sub("   ", '', str(s))
    s = re.sub("(  )",'.', str(s))
    modified_sentences = []
    sentences = s.split(".")
    for i in sentences:
        i = i.strip()
        if len(i) != 0:
            i += '.'
            modified_sentences.append(i)
    s = " ". join(modified_sentences)
    words = []
    for i in s.split(" "):
        words.append(contractions.fix(i))
    s = " ".join(words)
    words = []
    s = re.sub(r"[<>()|&©ø\[\]\'\",;?~*!]", ' ', str(s)).lower()
    for i in s.split(" "):
        if i not in engStopwords and i not in words:
            words.append(i)
    s = " ".join(words)
    s = re.sub(' +', ' ', s)
    return s

In [38]:
preprocessingRow(df1['content'].iloc[369])

'2022 came celebratory end bts especially member j-hope new york making another fabulous display skills. away members seemed enjoyed fullest solo performance dick clark’s year’s rockin’ eve live stage 3 songs track ‘= equal sign ’ ‘chicken noodle soup’ collaboration becky g bts’ ‘butter’ holiday remix . officially became second south korean soloist perform event following psy. also j-hope’s third time group stages 2017 2019. live. returning times square performed penultimate act alongside multiple singers around world spoke ryan seacrest host show back hotel room turned broadcast speak fans. congratulated sparkling perfectionist expressed sadness unable use voice talked slipping incident due rainy weather rehearsal stage. members’ year wishes. soon 12 began korea jungkook jimin v took fan community platform weverse share wishes fans well discuss plans coming year. kept brief wishing successful happy ahead wrote big letter feelings seemingly bottled long release music meeting composers.

### Although, I was not able to figure out how to remove a link because, I wasn't able to find any regular expression for removing it. I did research on link removal fuctions through BeautifulSoup but wasnt able to do it.

### The above function will work perfectly when provided row by row input but at that time we cannot utilize the multiprocessing functions of NLP. Hence creating same function, but this time we shall pass an entire column rather than a row to function

In [39]:
from tqdm.auto import tqdm

In [40]:
def preprocessingColumn(col):
    for row in tqdm(col, total=col.shape[0]):
        cleanedData = preprocessingRow(row)
        yield cleanedData

In [41]:
preprocessingColumn(df1['content']) # will result in a Generator object

<generator object preprocessingColumn at 0x000001A464E8D0B0>

# Storing cleaned data in the dataframe

In [42]:
cleanedContent = preprocessingColumn(df1['content'])

### Using spacy to fasten the process of Data Pre-processing

In [43]:
import spacy
nlp = spacy.load('en_core_web_sm', disable=['ner', 'parser'])

### Perorming Data Pre-processing on the Dataframe's content column

In [44]:
cleanedText = [str(doc) for doc in nlp.pipe(cleanedContent, batch_size=5000, n_process=-1)]

  0%|          | 0/805 [00:00<?, ?it/s]

In [45]:
cleanedText[0]

'reaching hotel city rm revealed stay would four days added step dinner. sat roadside open-air restaurant feasted beer burgers fries. said starving right now. grab food. much quieter expected feels like rural town. familiar atmosphere. attended art basel explained camera details fair. also gave glimpse noodles followed soup wrap. showing pattern ping pong table looks bts symbol. spoke pieces viewed them. took tram visit foundation beyeler museum. later walk city. third day visited kunstmuseum vitra design museum gallery. walked around showed chair fans breaking news guys. coldplay chris martin made displayed see give call. amazing. next lucerne hiked mount rigi. recalling previous remember crossing bridge buying souvenirs. reminded bon voyage reality show featuring members jin suga j-hope jimin v jungkook. speaking rode ssb train boat mountain track road cable cars planning go ride again. travel switzerland ended tinguely. flew paris attend pinault collection musee orsay. went centre g

In [46]:
cleanedText[369]

'2022 came celebratory end bts especially member j-hope new york making another fabulous display skills. away members seemed enjoyed fullest solo performance dick clark’s year’s rockin’ eve live stage 3 songs track ‘= equal sign ’ ‘chicken noodle soup’ collaboration becky g bts’ ‘butter’ holiday remix . officially became second south korean soloist perform event following psy. also j-hope’s third time group stages 2017 2019. live. returning times square performed penultimate act alongside multiple singers around world spoke ryan seacrest host show back hotel room turned broadcast speak fans. congratulated sparkling perfectionist expressed sadness unable use voice talked slipping incident due rainy weather rehearsal stage. members’ year wishes. soon 12 began korea jungkook jimin v took fan community platform weverse share wishes fans well discuss plans coming year. kept brief wishing successful happy ahead wrote big letter feelings seemingly bottled long release music meeting composers.

In [47]:
len(cleanedText)

805

### Checking for any errors in the process of data cleaning, such that out code might have skipped processing an entire row.

In [48]:
flag = 1
for i in range(len(cleanedText)):
    if len(cleanedText[i]) == 0:
        print(f"Error at: {i}")
        flag = 0
if flag == 1:
    print("No error, in process of Data pre-processing")

No error, in process of Data pre-processing


# Trying to summarize one row

In [49]:
from gensim.models import Word2Vec
from scipy import spatial
import networkx as nx

In [50]:
text = initial_preprocessing(df1['content'].iloc[405])
text

'BTS member Jimin\'s first-ever OST, With You, for the K-drama Our Blues was released on Sunday. The track also features Jimin\'s friend and former HOTSHOT and Want To One member Ha Sung Woon. YamYam Entertainment released the music video on their YouTube channel. The song is part of episode six of the K-drama. (Also Read | BTS: Jimin\'s Our Blues OST release date changed; his friend Ha Sung Woon reveals details about their upcoming song)The song features actors Shin Min-a and Lee Byung-hun in Jeju Island. The lyrics of the song start with Jimin singing, "I want to be with you and I want to stay with you, Just like the stars shining bright." Ha Sung Woon starts singing after Jimin. The duo sing in both Korean as well as English. A part of Jimin\'s lyrics also goes, "I am nothing without your love, I promise I will never leave your love my heart is beating because of you."Reacting to the clip, BTS fandom ARMY dropped comments praising Jimin, Ha Sung Woon and the song. A person wrote, "P

In [51]:
sentences = [i.strip()+"." for i in text.split(".") if len(i) != 0]
sentences

["BTS member Jimin's first-ever OST, With You, for the K-drama Our Blues was released on Sunday.",
 "The track also features Jimin's friend and former HOTSHOT and Want To One member Ha Sung Woon.",
 'YamYam Entertainment released the music video on their YouTube channel.',
 'The song is part of episode six of the K-drama.',
 "(Also Read | BTS: Jimin's Our Blues OST release date changed; his friend Ha Sung Woon reveals details about their upcoming song)The song features actors Shin Min-a and Lee Byung-hun in Jeju Island.",
 'The lyrics of the song start with Jimin singing, "I want to be with you and I want to stay with you, Just like the stars shining bright.',
 '" Ha Sung Woon starts singing after Jimin.',
 'The duo sing in both Korean as well as English.',
 'A part of Jimin\'s lyrics also goes, "I am nothing without your love, I promise I will never leave your love my heart is beating because of you.',
 '"Reacting to the clip, BTS fandom ARMY dropped comments praising Jimin, Ha Sung W

In [52]:
sentence_tokens = []
for sentence in sentences:
    for word in preprocessingRow(sentence).split(" "):
        if word not in sentence_tokens:
            sentence_tokens.append(word)

In [53]:
w2v = Word2Vec(sentence_tokens, vector_size=1, min_count = 1, epochs = 1000)
sentence_embeddings = [[w2v.wv[word][0] for word in words] for words in sentence_tokens]
max_len = max([len(tokens) for tokens in sentence_tokens])
sentence_embeddings = [np.pad(embedding,(0,max_len-len(embedding)),'constant') for embedding in sentence_embeddings]

In [54]:
similarity_matrix = np.zeros([len(sentence_tokens), len(sentence_tokens)])
for i,row_embedding in enumerate(sentence_embeddings):
    for j,column_embedding in enumerate(sentence_embeddings):
        similarity_matrix[i][j]=1-spatial.distance.cosine(row_embedding,column_embedding)

In [55]:
nx_graph = nx.from_numpy_array(similarity_matrix)
scores = nx.pagerank(nx_graph)

In [56]:
total_summary_sentences = len(sentences) // 4
top_sentence={sentence:scores[index] for index,sentence in enumerate(sentences)}
top=dict(sorted(top_sentence.items(), key=lambda x: x[1], reverse=True)[:total_summary_sentences])

In [57]:
summary = ""
for sentence in sentences:
    if sentence in top.keys():
        summary += sentence
print(summary)

The track also features Jimin's friend and former HOTSHOT and Want To One member Ha Sung Woon.YamYam Entertainment released the music video on their YouTube channel.You are special and indescribable! You deserve the world." "Sung Woon and Jimin's voice together is such a masterpiece," said a fan."Lyrics are just exactly what Jimin will say to ARMYs #WithYouJiminOST #WithYou @BTS_twt.Jimin you are so lovely, we so lovely, lovely, lovely.During a recent live session with J-Hope and V, Jimin had replied to a fan who asked about his OST for the series.While V sang Christmas Tree for Our Beloved Summer, Jin released Yours for Jirisan.


### For below text

In [58]:
text

'BTS member Jimin\'s first-ever OST, With You, for the K-drama Our Blues was released on Sunday. The track also features Jimin\'s friend and former HOTSHOT and Want To One member Ha Sung Woon. YamYam Entertainment released the music video on their YouTube channel. The song is part of episode six of the K-drama. (Also Read | BTS: Jimin\'s Our Blues OST release date changed; his friend Ha Sung Woon reveals details about their upcoming song)The song features actors Shin Min-a and Lee Byung-hun in Jeju Island. The lyrics of the song start with Jimin singing, "I want to be with you and I want to stay with you, Just like the stars shining bright." Ha Sung Woon starts singing after Jimin. The duo sing in both Korean as well as English. A part of Jimin\'s lyrics also goes, "I am nothing without your love, I promise I will never leave your love my heart is beating because of you."Reacting to the clip, BTS fandom ARMY dropped comments praising Jimin, Ha Sung Woon and the song. A person wrote, "P

### We got summary as

In [59]:
summary

'The track also features Jimin\'s friend and former HOTSHOT and Want To One member Ha Sung Woon.YamYam Entertainment released the music video on their YouTube channel.You are special and indescribable! You deserve the world." "Sung Woon and Jimin\'s voice together is such a masterpiece," said a fan."Lyrics are just exactly what Jimin will say to ARMYs #WithYouJiminOST #WithYou @BTS_twt.Jimin you are so lovely, we so lovely, lovely, lovely.During a recent live session with J-Hope and V, Jimin had replied to a fan who asked about his OST for the series.While V sang Christmas Tree for Our Beloved Summer, Jin released Yours for Jirisan.'

# Time to implement it on actual Dataframe

In [60]:
def summarizerRow(s):
    # Applying only the necessary Pre-processing
    text = initial_preprocessing(s)
    # print("1")
    
    # Generating the tokens
    sentences = [i.strip()+"." for i in text.split(".") if len(i) != 0]
    sentence_tokens = [[word for word in preprocessingRow(sentence).split(" ")] for sentence in sentences]
    # print("2")
    
    # Calculating the embedding for each word
    w2v = Word2Vec(sentence_tokens, vector_size=1, min_count = 1, epochs = 1000)
    sentence_embeddings = [[w2v.wv[word][0] for word in words] for words in sentence_tokens]
    max_len = max([len(tokens) for tokens in sentence_tokens])
    sentence_embeddings = [np.pad(embedding,(0,max_len-len(embedding)),'constant') for embedding in sentence_embeddings]
    # print("3")
    
    # Developing the similarity matrix based on cosine similarity
    similarity_matrix = np.zeros([len(sentence_tokens), len(sentence_tokens)])
    for i,row_embedding in enumerate(sentence_embeddings):
        for j,column_embedding in enumerate(sentence_embeddings):
            similarity_matrix[i][j]=1-spatial.distance.cosine(row_embedding,column_embedding)
    # print("4")
    
    # Implementing the TextRank
    nx_graph = nx.from_numpy_array(similarity_matrix)
    try:
        scores = nx.pagerank(nx_graph,  max_iter=100000, tol=1.0e-2)

        # Sorting out the importing sentences
        total_summary_sentences = len(sentences) // 3
        top_sentence={sentence:scores[index] for index,sentence in enumerate(sentences)}
        top=dict(sorted(top_sentence.items(), key=lambda x: x[1], reverse=True)[:total_summary_sentences])

        # Joining the most important sentences
        summary = ""
        for sentence in sentences:
            if sentence in top.keys():
                summary += sentence + " "
        # print("5")
        return summary
    except:
        summary = ""
        # print("6")
        return summary

In [61]:
print(summarizerRow(df1['content'].iloc[402]))

At the centre of this plan stand the many official YouTube channels that partake in building a community on the platform for them. This is led primarily by BANGTANTV, the group’s behind-the-scenes deal breaker and content carrier. The biggest contributor to these channels continues to be boy group BTS that has now achieved a massive milestone. 7 billion views across all their official channels, BTS has become the most viewed artist in the history of the platform. The difference between the first and the tenth acts on the list, BTS and Ariana Grande, is over 6 billion views. 


### Developing a similar fuction column-wise

In [62]:
def summarizerColumn(col):
    for row in tqdm(col, total=col.shape[0]):
        summary = summarizerRow(row)
        yield summary

In [63]:
summaryGenerator = summarizerColumn(df1['content'])

In [64]:
summary = [str(doc) for doc in nlp.pipe(summaryGenerator, batch_size=5000, n_process=-1)]

  0%|          | 0/805 [00:00<?, ?it/s]

In [65]:
summary[0]

'As he sat at a roadside open-air restaurant, RM feasted on beer, burgers and fries. " RM attended Art Basel and explained on camera the details of the art fair. He also gave a glimpse as he had noodles and beer which was followed by soup noodles and wrap. After that, RM took a tram to visit the Foundation Beyeler, a museum. As he walked around, RM showed a chair to his fans and said, "I have breaking news for you guys. Coldplay\'s Chris Martin made a chair and it is displayed in the Vitra Design Museum. " RM next visited Lucerne and hiked to Mount Rigi. " RM\'s travel in Switzerland ended with a visit to the Museum Tinguely. '

In [66]:
df1['content'].shape

(805,)

In [67]:
len(summary)

805

In [68]:
df1['content'].iloc[804]

'After a lot of teasing, Benny Blanco’s collaborative song with BTS and Snoop Dogg is finally here! ‘Bad Decisions’ features BTS’ vocal line members Jin, Jimin, V and Jungkook along with rapper Snoop Dogg, and is a pre-release single from Benny Blanco’s upcoming full-length album \xa0Released on August 5 at 9:30 am IST, the light beat of ‘Bad Decisions’ becomes the perfect background for its lyrics that express honest feelings to a loved person  The dance track carries a cool feeling, which helps the listener feel refreshed on a hot summer day Meanwhile, the hilarious music video sees Benny Blanco getting ready to attend a BTS concert, only to find out that he has got the date wrong, and to be told that he\'s made a "bad decision"!Check out the fun music video for ‘Bad Decisions’, below:\xa0https://www youtube com/embed/BGNkkVrJZksPrior to the release, two music video teasers were dropped, adopting a style reminiscent of that of a movie trailer  With a classic narration accompanying BT

In [69]:
summary[804]

'With a classic narration accompanying BTS’ shots from their previous music video, the teasers hilariously teased the then-upcoming song, without actually giving away any hint about the song itself or its music video Further, Benny Blanco also kicked off a hunt for ‘Bad Decisions’, by sharing a video which showed him hiding a USB drive containing the song “somewhere in LA”. With an hour and a half to go before the song’s release, Benny Blanco took to Twitter again, this time dropping an address for the USB drive’s location, and asking fans to meet up there!From the teasers to a hunt for a USB drive and more, ‘Bad Decisions’ has been a series of chaotically fun events! Following today’s music video premiere and release of ‘Bad Decisions’ on streaming platforms, the lyric video of the song will be out on BTS’ YouTube channel tomorrow, at 9:30 am IST on August 6. '

In [70]:
df2 = pd.DataFrame()

In [71]:
df2['Original Content'] = df1['content'].apply(initial_preprocessing)
df2.head()

Original Content
0  After reaching his hotel in the city, RM revea...
1  RM aka Kim Namjoon was the first member to joi...
2  Billie Eilish's concert was held in Seoul, Sou...
3  BTS ARMY you all would be missing the members ...
4  BTS member Kim Seokjin aka Jin has the capacit...

In [72]:
df2.shape

(805, 1)

In [73]:
df2['Original Content'].iloc[369]

'2022 came to a celebratory end for BTS, especially member J-Hope, who was in New York, making another fabulous display of his skills.While away from the members, he seemed to have enjoyed it to the fullest with a solo performance at Dick Clark’s New Year’s Rockin’ Eve where he did a live stage of 3 songs, solo track ‘= (Equal Sign)’, ‘Chicken Noodle Soup’ his collaboration track with Becky G, and BTS’ ‘Butter’ (Holiday Remix).He officially became only the second South Korean soloist to perform at the event, following PSY.This was also J-Hope’s third time at Dick Clark’s New Year’s Rockin’ Eve after group stages with BTS in 2017 and 2019 .J-Hope’s live.After returning from Times Square where he performed as the penultimate act alongside multiple other singers from around the world and spoke to Ryan Seacrest, the host of the show, J-Hope was back at his hotel room and turned on a live broadcast to speak with his fans.As they congratulated him on his sparkling performance, being the perf

In [74]:
df2['New Content'] = summary
df2.head()

Original Content  \
0  After reaching his hotel in the city, RM revea...   
1  RM aka Kim Namjoon was the first member to joi...   
2  Billie Eilish's concert was held in Seoul, Sou...   
3  BTS ARMY you all would be missing the members ...   
4  BTS member Kim Seokjin aka Jin has the capacit...   

                                         New Content  
0  As he sat at a roadside open-air restaurant, R...  
1  RM aka Kim Namjoon was the first member to joi...  
2  They really enjoyed the concert as the audienc...  
3  The boys are going to complete their projects ...  
4  Some days back, we saw the Sea of Jin concept ...

In [75]:
df2.drop(df2[df2['New Content'] == ''].index[0], axis=0, inplace=True)

In [76]:
df2[df2['New Content'] == '']

Empty DataFrame
Columns: [Original Content, New Content]
Index: []

In [77]:
df2.isna().sum()

Original Content    0
New Content         0
dtype: int64

In [78]:
df2.shape

(804, 2)

In [79]:
df2[df2['New Content'].isna()]

Empty DataFrame
Columns: [Original Content, New Content]
Index: []

In [80]:
try:
    df2.to_csv('./Summarized News.csv', index=False)
    print("Your file has been created !!!")
except:
    print("Failed to create fail. Might be due to permission related errors.")

Your file has been created !!!


# Finally working on our Summarized File

In [81]:
df3 = pd.read_csv('./Summarized News.csv')
df3.head()

Original Content  \
0  After reaching his hotel in the city, RM revea...   
1  RM aka Kim Namjoon was the first member to joi...   
2  Billie Eilish's concert was held in Seoul, Sou...   
3  BTS ARMY you all would be missing the members ...   
4  BTS member Kim Seokjin aka Jin has the capacit...   

                                         New Content  
0  As he sat at a roadside open-air restaurant, R...  
1  RM aka Kim Namjoon was the first member to joi...  
2  They really enjoyed the concert as the audienc...  
3  The boys are going to complete their projects ...  
4  Some days back, we saw the Sea of Jin concept ...

In [82]:
df3.isna().sum()

Original Content    0
New Content         0
dtype: int64

In [83]:
content = df3['Original Content'].iloc[370]
content

"On January 23, JYP Entertainment released new concept images for Stray Kids’ second fan meeting ‘SKZ’s Chocolate Factory’.Each member looks amazing in the soft pastel clothes and pretty accessories as they work in the chocolate factory.The fan meeting will be held on February 12 and 13.On February 13th, an offline fan meeting and an online paid live broadcast on the Beyond Live platform will be held at the same time, and precious memories will be made with domestic and foreign fans.Tickets for offline performances were pre-purchased for the fan club from 8:00 pm to 11:59 pm on January 17th for members of the 2nd period of the official fan club STAY, and all seats were sold out at the same time as they opened.Thanks to such enthusiastic support, JYP Entertainment opened additional seats available for viewing at 8 pm on January 19th, and this also sold out quickly, realizing the power of Stray Kids' tickets.This fan meeting is the first in about a year since the first official fan meeti

In [84]:
summary = df3['New Content'].iloc[370]
summary

"Tickets for offline performances were pre-purchased for the fan club from 8:00 pm to 11:59 pm on January 17th for members of the 2nd period of the official fan club STAY, and all seats were sold out at the same time as they opened. Thanks to such enthusiastic support, JYP Entertainment opened additional seats available for viewing at 8 pm on January 19th, and this also sold out quickly, realizing the power of Stray Kids' tickets. At the '2021 The Fact Music Awards', '2021 Asian Artist Awards', and '2021 Mnet Asian Music Awards' respectively, they won the Artist of the Year Award, Performance Award of the Year, and Worldwide Fans' Choice Top Ten respectively. "

In [85]:
content_split = []
for i in content.split("."):
    i = i.strip()
    if len(i) != 0 and i != " ":
        content_split.append(i)
len(content_split)

10

In [86]:
summary_split = []
for i in summary.split("."):
    i = i.strip()
    if len(i) != 0 and i != " ":
        summary_split.append(i)
len(summary_split)

3

In [87]:
removed_lines = []
for i in content_split:
    if i not in summary_split:
        removed_lines.append(i)
len(removed_lines)

7

### Combining all above process into a function

In [88]:
def removed_lines(content, summary):
    content_split = []
    for i in content.split("."):
        i = i.strip()
        if len(i) != 0 and i != " ":
            content_split.append(i)
    
    summary_split = []      
    for i in summary.split("."):
        i = i.strip()
        if len(i) != 0 and i != " ":
            summary_split.append(i)
    
    removed_line = []
    for i in content_split:
        if i not in summary_split:
            removed_line.append(i)
            
    removedLines = ". ".join(removed_line) 
    return removedLines

In [89]:
removed_line = df3.apply(lambda x: removed_lines(x['Original Content'], x['New Content']), axis=1)

In [90]:
removed_line

0      After reaching his hotel in the city, RM revea...
1      The group released their debut single album 2 ...
2      Billie Eilish's concert was held in Seoul, Sou...
3      BTS ARMY you all would be missing the members ...
4      BTS member Kim Seokjin aka Jin has the capacit...
                             ...                        
799    BTS has conquered the world with their group r...
800    Since it was released, the meaningful song rec...
801    After checking out his live video, BTS' Jungko...
802    BTS' eldest member Jin has shared pictures and...
803    After a lot of teasing, Benny Blanco’s collabo...
Length: 804, dtype: object

In [91]:
len(removed_line)

804

In [92]:
df3.shape

(804, 2)

In [93]:
df3['Removed Lines'] = removed_line
df3.head()

Original Content  \
0  After reaching his hotel in the city, RM revea...   
1  RM aka Kim Namjoon was the first member to joi...   
2  Billie Eilish's concert was held in Seoul, Sou...   
3  BTS ARMY you all would be missing the members ...   
4  BTS member Kim Seokjin aka Jin has the capacit...   

                                         New Content  \
0  As he sat at a roadside open-air restaurant, R...   
1  RM aka Kim Namjoon was the first member to joi...   
2  They really enjoyed the concert as the audienc...   
3  The boys are going to complete their projects ...   
4  Some days back, we saw the Sea of Jin concept ...   

                                       Removed Lines  
0  After reaching his hotel in the city, RM revea...  
1  The group released their debut single album 2 ...  
2  Billie Eilish's concert was held in Seoul, Sou...  
3  BTS ARMY you all would be missing the members ...  
4  BTS member Kim Seokjin aka Jin has the capacit...

In [94]:
len(df3['Original Content'].iloc[0].split("."))-len(df3['New Content'].iloc[0].split("."))

17

In [95]:
len(df3['Removed Lines'].iloc[0].split("."))

17

In [96]:
from sentence_transformers import SentenceTransformer, util

In [97]:
model = SentenceTransformer('all-MiniLM-L6-v2')

In [98]:
originalC = df3['Original Content'].iloc[1]
newC = df3['New Content'].iloc[1]

In [99]:
en_1 = model.encode(originalC)
en_2 = model.encode(newC)
print(type(en_1))

<class 'numpy.ndarray'>


### Cosine Similarity

In [100]:
result = util.cos_sim(en_1, en_2)

print(float(result))
print(type(result))

0.7904312014579773
<class 'torch.Tensor'>


In [101]:
result_float = result.item()
print(result_float)
print(type(result_float))

0.7904312014579773
<class 'float'>


In [102]:
def metrics(originalC, newC):
    en_1 = model.encode(originalC)
    en_2 = model.encode(newC)
    result = util.cos_sim(en_1, en_2)
    # print(float(result))
    return float(result)

In [103]:
cosine = df3.apply(lambda x: metrics(x['Original Content'], x['New Content']), axis=1)

In [104]:
cosine.shape

(804,)

In [105]:
df3['Cosine Similarity'] = cosine

In [106]:
df3.head()

Original Content  \
0  After reaching his hotel in the city, RM revea...   
1  RM aka Kim Namjoon was the first member to joi...   
2  Billie Eilish's concert was held in Seoul, Sou...   
3  BTS ARMY you all would be missing the members ...   
4  BTS member Kim Seokjin aka Jin has the capacit...   

                                         New Content  \
0  As he sat at a roadside open-air restaurant, R...   
1  RM aka Kim Namjoon was the first member to joi...   
2  They really enjoyed the concert as the audienc...   
3  The boys are going to complete their projects ...   
4  Some days back, we saw the Sea of Jin concept ...   

                                       Removed Lines  Cosine Similarity  
0  After reaching his hotel in the city, RM revea...           0.796752  
1  The group released their debut single album 2 ...           0.790431  
2  Billie Eilish's concert was held in Seoul, Sou...           0.840863  
3  BTS ARMY you all would be missing the members ...           0.709553  
4  BTS member Kim Seokjin aka Jin has the capacit...           0.719757

In [107]:
min(cosine)

0.36477231979370117

In [108]:
max(cosine)

0.9571455717086792

In [109]:
np.mean(cosine)

0.7547245741409448

In [110]:
np.median(cosine)

0.7777727246284485

In [111]:
nlp1 = spacy.load("en_core_web_lg")

In [112]:
doc1 = nlp(originalC)
doc2 = nlp(newC)
print(doc1.similarity(doc2)) 

0.9229698009279503


In [113]:
def nlptextsimilarity(originalC, newC):
    doc1 = nlp(originalC)
    doc2 = nlp(newC)
    return doc1.similarity(doc2) 

In [114]:
nlptextsimi = df3.apply(lambda x: nlptextsimilarity(x['Original Content'], x['New Content']), axis=1)

In [115]:
df3['NLP Text Similarity'] = nlptextsimi

In [116]:
df3.head()

Original Content  \
0  After reaching his hotel in the city, RM revea...   
1  RM aka Kim Namjoon was the first member to joi...   
2  Billie Eilish's concert was held in Seoul, Sou...   
3  BTS ARMY you all would be missing the members ...   
4  BTS member Kim Seokjin aka Jin has the capacit...   

                                         New Content  \
0  As he sat at a roadside open-air restaurant, R...   
1  RM aka Kim Namjoon was the first member to joi...   
2  They really enjoyed the concert as the audienc...   
3  The boys are going to complete their projects ...   
4  Some days back, we saw the Sea of Jin concept ...   

                                       Removed Lines  Cosine Similarity  \
0  After reaching his hotel in the city, RM revea...           0.796752   
1  The group released their debut single album 2 ...           0.790431   
2  Billie Eilish's concert was held in Seoul, Sou...           0.840863   
3  BTS ARMY you all would be missing the members ...           0.709553   
4  BTS member Kim Seokjin aka Jin has the capacit...           0.719757   

   NLP Text Similarity  
0             0.947731  
1             0.922970  
2             0.938945  
3             0.923807  
4             0.909008

In [117]:
min(nlptextsimi)

0.5577889218703898

In [118]:
max(nlptextsimi)

0.9906746332886787

In [119]:
np.mean(nlptextsimi)

0.9239102736903291

In [120]:
np.median(nlptextsimi)

0.9329814334145199

In [121]:
def harmonicMean(x, y):
    return (2*x*y)/(x+y)

In [122]:
harmean = df3.apply(lambda x: harmonicMean(x['Cosine Similarity'], x['NLP Text Similarity']), axis=1)

In [123]:
df3['Harmonic Mean'] = harmean

In [124]:
df3.head()

Original Content  \
0  After reaching his hotel in the city, RM revea...   
1  RM aka Kim Namjoon was the first member to joi...   
2  Billie Eilish's concert was held in Seoul, Sou...   
3  BTS ARMY you all would be missing the members ...   
4  BTS member Kim Seokjin aka Jin has the capacit...   

                                         New Content  \
0  As he sat at a roadside open-air restaurant, R...   
1  RM aka Kim Namjoon was the first member to joi...   
2  They really enjoyed the concert as the audienc...   
3  The boys are going to complete their projects ...   
4  Some days back, we saw the Sea of Jin concept ...   

                                       Removed Lines  Cosine Similarity  \
0  After reaching his hotel in the city, RM revea...           0.796752   
1  The group released their debut single album 2 ...           0.790431   
2  Billie Eilish's concert was held in Seoul, Sou...           0.840863   
3  BTS ARMY you all would be missing the members ...           0.709553   
4  BTS member Kim Seokjin aka Jin has the capacit...           0.719757   

   NLP Text Similarity  Harmonic Mean  
0             0.947731       0.865708  
1             0.922970       0.851574  
2             0.938945       0.887202  
3             0.923807       0.802628  
4             0.909008       0.803388

In [125]:
df3.to_csv('./Final Result.csv')

In [126]:
min(harmean)

0.49884301558564753

In [127]:
max(harmean)

0.9623399165870538

In [128]:
np.mean(harmean)

0.8265396330480197

In [129]:
np.median(harmean)

0.8414175524334191